concent_api/core/payments/report.py

```diff
@@ -0,0 +1,753 @@
+
+        task_to_compute = deserialize_message(subtask.task_to_compute.data.tobytes())
+
+        related_payments_for_pair = get_related_payments_for_pair(
+            task_to_compute.requestor_ethereum_address,
+            task_to_compute.provider_ethereum_address,
+        )
+
+        (errors, warnings) = get_errors_and_warnings(
+            related_payments_for_pair,
+            related_pending_responses_for_pair,
+        )
+
+        pair_report = PairReport(
+            requestor=subtask.requestor,
+            provider=subtask.provider,
+            pending_responses=related_pending_responses_for_pair,
+            payments=related_payments_for_pair,
+            errors=errors,
+            warnings=warnings,
+            summary=create_pair_summary(
+                requestor=subtask.requestor,
+                provider=subtask.provider,
+                pending_responses=related_pending_responses_for_pair,
+                payments=related_payments_for_pair,
+            ),
+        )
+        pair_reports.append(pair_report)
+
+    return Report(
+        pair_reports,
+        summary=create_report_summary(pair_reports),
+    )
+
+
+def get_subtasks_with_distinct_clients_pairs() -> QuerySet:
+    """
+    Get Subtasks with distinct pairs of requestor and provider Clients.
+    """
+    return Subtask.objects.distinct('provider', 'requestor')
+
+
+def get_related_pending_responses_for_pair(requestor: Client, provider: Client) -> PendingResponses:
+    """ Group and return pending responses from Subtask queryset. """
+    subtask_ids_for_pair = Subtask.objects.filter(  # pylint: disable=no-member
+        requestor=requestor,
+        provider=provider,
+    ).values_list('subtask_id', flat=True)
+    return PendingResponses(
+        subtask_results_settled=PendingResponse.objects.filter(
+            response_type=PendingResponse.ResponseType.SubtaskResultsSettled.name,  # pylint: disable=no-member
+            subtask__subtask_id__in=subtask_ids_for_pair,
+        ),
+        subtask_results_rejected=PendingResponse.objects.filter(
+            response_type=PendingResponse.ResponseType.SubtaskResultsRejected.name,  # pylint: disable=no-member
+            subtask__subtask_id__in=subtask_ids_for_pair,
+        ),
+        force_payment_committed=PendingResponse.objects.filter(
+            response_type=PendingResponse.ResponseType.ForcePaymentCommitted.name,  # pylint: disable=no-member
+            subtask__subtask_id__in=subtask_ids_for_pair,
+        ),
+    )
+
+
+def get_related_payments_for_pair(requestor_ethereum_address: str, provider_ethereum_address: str) -> Payments:
+    """ Get all payments between given requestor/provider pair. """
+    regular_payments = service.get_list_of_payments(  # pylint: disable=no-value-for-parameter
+        requestor_eth_address=requestor_ethereum_address,
+        provider_eth_address=provider_ethereum_address,
+        min_block_timestamp=get_oldest_subtask_creation_date(),
```

Comment:
Using creation time of the oldest subtask is just a heuristic. It's not foolproof as you have just noted. The idea is to prevent stuff from before the last database reset from getting in the way - so we don't want to go too far in the past.

But yeah, using the timestamp from RCT seems like a better heuristic. It's still not that far in the past but would eliminate the problems your're talking about.